File: test_nodl_generator_cpp/test/test_pubsub.cpp
#include <chrono>
#include <memory>
#include <string>

#include <gtest/gtest.h>
#include <rclcpp/rclcpp.hpp>
#include <std_msgs/msg/string.hpp>

#include "pubsub_node_derived.hpp"

using namespace std::chrono_literals;

class PubsubRoundTrip : public ::testing::Test
{
protected:
  void SetUp() override { rclcpp::init(0, nullptr); }
  void TearDown() override { rclcpp::shutdown(); }
};

// End-to-end: harness publishes on /echo_in, derived node receives via the
// generated subscription, republishes through the generated publisher on
// /echo_out, harness subscribes and confirms the payload round-tripped.
TEST_F(PubsubRoundTrip, EchoesThroughGeneratedHandles)
{
  auto node = std::make_shared<PubsubNode>();
  auto harness = std::make_shared<rclcpp::Node>("test_harness");

  auto qos = rclcpp::QoS(10).reliable();

  std::string received;
  auto sub = harness->create_subscription<std_msgs::msg::String>(
    "/echo_out", qos,
    [&received](std_msgs::msg::String::ConstSharedPtr msg) { received = msg->data; });

  auto pub = harness->create_publisher<std_msgs::msg::String>("/echo_in", qos);

  rclcpp::executors::SingleThreadedExecutor exec;
  exec.add_node(node);
  exec.add_node(harness);

  // Wait for pub/sub discovery — bounded.
  const auto deadline = std::chrono::steady_clock::now() + 2s;
  while (std::chrono::steady_clock::now() < deadline &&
         (pub->get_subscription_count() == 0 || sub->get_publisher_count() == 0))
  {
    exec.spin_some();
    std::this_thread::sleep_for(10ms);
  }
  ASSERT_GT(pub->get_subscription_count(), 0u) << "derived node didn't subscribe to /echo_in";
  ASSERT_GT(sub->get_publisher_count(), 0u) << "derived node didn't publish on /echo_out";

  std_msgs::msg::String msg;
  msg.data = "hello";
  pub->publish(msg);

  const auto recv_deadline = std::chrono::steady_clock::now() + 2s;
  while (std::chrono::steady_clock::now() < recv_deadline && received.empty()) {
    exec.spin_some();
    std::this_thread::sleep_for(10ms);
  }

  EXPECT_EQ(node->receive_count, 1);
  EXPECT_EQ(node->last_received, "hello");
  EXPECT_EQ(received, "echo: hello");
}

// Cheap structural checks — useful as a quick smoke signal if the round-trip
// is failing for environment reasons (no DDS discovery, etc.).
TEST_F(PubsubRoundTrip, GeneratedHandlesExist)
{
  auto node = std::make_shared<PubsubNode>();
  EXPECT_STREQ(node->get_name(), "pubsub_node");
  // pub_echo_out_ is protected on the base; visible here because PubsubNode
  // inherits and we're testing through the derived type's interface.
}

File: test_nodl_generator_py/test/test_pubsub.py
# SPDX-FileCopyrightText: 2026 Open Source Robotics Foundation, Inc.
# SPDX-License-Identifier: Apache-2.0
"""End-to-end pub/sub test for the exported generator macro."""

import time

import pytest
import rclpy
from rclpy.executors import SingleThreadedExecutor
from rclpy.node import Node
from std_msgs.msg import String

from test_nodl_generator_py.pubsub_node_impl import PubsubNode


@pytest.fixture
def ros_context():
    rclpy.init()
    try:
        yield
    finally:
        rclpy.shutdown()


def _spin_until(executor, predicate, timeout=2.0):
    deadline = time.monotonic() + timeout
    while time.monotonic() < deadline and not predicate():
        executor.spin_once(timeout_sec=0.05)
    return predicate()


def test_echoes_through_generated_handles(ros_context):
    node = PubsubNode()
    harness = Node('test_harness')

    received = []
    harness.create_subscription(String, '/echo_out', lambda m: received.append(m.data), 10)
    pub = harness.create_publisher(String, '/echo_in', 10)

    executor = SingleThreadedExecutor()
    executor.add_node(node)
    executor.add_node(harness)
    try:
        assert node.get_name() == 'pubsub_node'
        assert hasattr(node, 'pub_echo_out')
        assert hasattr(node, 'sub_echo_in')
        assert _spin_until(executor, lambda: pub.get_subscription_count() > 0)

        pub.publish(String(data='hello'))
        assert _spin_until(executor, lambda: bool(received))
        assert received == ['echo: hello']
    finally:
        executor.shutdown()
        node.destroy_node()
        harness.destroy_node()
